taurus/util/tests/test_foreach.py

```diff
@@ -0,0 +1,15 @@
+from taurus.entity.object import ProcessRun, MaterialRun, IngredientRun
+from taurus.util.impl import recursive_foreach
+
+
+def test_recursive_foreach():
```

Comment:
This test needs to do a full process->material->ingredient->process->material set in order to actually guarantee behavior.

Reply:
Why do you say that?  If there's something to add, I'd guess that it should be a measurement rather than another process.

Reply:
Agreed (both that an additional process->material link is not necessary, and that a measurement should be added)

Reply:
Done, plus a property template for good measure.

Reply:
I can understand pushback on the final process, but if you do not at least have 1 full cycle, you are not testing all references.

Reply:
Isn't material -> ingredient -> process -> material a full cycle?